Makefile

```diff
@@ -1,36 +1,118 @@
+endif
+
+install-golangci-lint:
+ifndef HAS_GOLANGCI_LINT
+	curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b $(shell go env GOPATH)/bin v1.26.0
+endif
 
-include scripts/inc.Makefile
+$(COMMANDS):
+	$(GO) build -ldflags "$(LDFLAGS)" -o $(BIN_OUTPUT_DIR)/$@ ./cmd/$@
 
-GOPKGS = ./pkg/adapter/... ./pkg/apis/... ./pkg/reconciler/...
+help: ## Display this help
+	@awk 'BEGIN {FS = ":.*?## "; printf "\n$(KREPO_DESC)\nUsage:\n  make \033[36m<source>\033[0m\n"} /^[a-zA-Z0-9._-]+:.*?## / {printf "  \033[36m%-20s\033[0m %s\n", $$1, $$2}' $(MAKEFILE_LIST)
 
-test:
-	$(GOTEST) -p=1 -cover -coverprofile=c.out $(GOPKGS)
+mod-download: ## Download go modules
+	$(GO) mod download
 
-coverage: ## Generate code coverage
-	$(GOTOOL) cover -html=c.out -o $(OUTPUT_DIR)$(PACKAGE)-coverage.html
+build: $(COMMANDS) ## Build the binary
 
-lint:
+release: ## Build release binaries
+	@set -e ; \
+	for bin in $(COMMANDS) ; do \
+		for platform in $(TARGETS); do \
+			GOOS=$${platform%/*} ; \
+			GOARCH=$${platform#*/} ; \
+			RELEASE_BINARY=$$bin-$${GOOS}-$${GOARCH} ; \
+			[ $${GOOS} = "windows" ] && RELEASE_BINARY=$${RELEASE_BINARY}.exe ; \
+			echo "GOOS=$${GOOS} GOARCH=$${GOARCH} $(GO) build -ldflags "$(LDFLAGS)" -o $(DIST_DIR)/$${RELEASE_BINARY}" ./cmd/$$bin ; \
+			GOOS=$${GOOS} GOARCH=$${GOARCH} $(GO) build -ldflags "$(LDFLAGS)" -o $(DIST_DIR)/$${RELEASE_BINARY} ./cmd/$$bin ; \
+		done ; \
+	done
+
+test: install-gotestsum ## Run unit tests
+	@mkdir -p $(TEST_OUTPUT_DIR)
+	$(GOTEST) -p=1 -race -cover -coverprofile=$(TEST_OUTPUT_DIR)/$(KREPO)-c.out $(GOPKGS)
+
+cover: test ## Generate code coverage
+	@mkdir -p $(COVER_OUTPUT_DIR)
+	$(GOTOOL) cover -html=$(TEST_OUTPUT_DIR)/$(KREPO)-c.out -o $(COVER_OUTPUT_DIR)/$(KREPO)-coverage.html
+
+lint: install-golangci-lint ## Lint source files
 	$(GOLINT) $(GOPKGS)
 
-fmt:
+fmt: ## Format source files
 	$(GOFMT) -s -w $(shell $(GO) list -f '{{$$d := .Dir}}{{range .GoFiles}}{{$$d}}/{{.}} {{end}} {{$$d := .Dir}}{{range .TestGoFiles}}{{$$d}}/{{.}} {{end}}' $(GOPKGS))
 
-fmt-test:
+fmt-test: ## Check source formatting
 	@test -z $(shell $(GOFMT) -l $(shell $(GO) list -f '{{$$d := .Dir}}{{range .GoFiles}}{{$$d}}/{{.}} {{end}} {{$$d := .Dir}}{{range .TestGoFiles}}{{$$d}}/{{.}} {{end}}' $(GOPKGS)))
 
-clean:
-	$(RM) -rf $(PACKAGE)-unit-tests.xml
-	$(RM) -rf c.out $(PACKAGE)-coverage.html
+IMAGES = $(foreach cmd,$(COMMANDS),$(cmd).image)
+images: $(IMAGES) ## Builds container images
+$(IMAGES): %.image:
+	$(DOCKER) build -t $(IMAGE_REPO)/$* -f ./cmd/$*/Dockerfile . ;
+
+CLOUDBUILD_TEST = $(foreach cmd,$(COMMANDS),$(cmd).cloudbuild-test)
+cloudbuild-test: $(CLOUDBUILD_TEST) ## Test container image build with Google Cloud Build
+$(CLOUDBUILD_TEST): %.cloudbuild-test:
+	gcloud builds submit $(BASE_DIR) --config cmd/$*/cloudbuild.yaml --substitutions COMMIT_SHA=${IMAGE_SHA},_KANIKO_IMAGE_TAG=_
```

Comment:
@sameersbn in the previous version of this goal you were setting the `_KANIKO_NO_PUSH` substitution to `true` instead of setting `_KANIKO_IMAGE_TAG` to `_`. In fact, the underscore does not prevent the push and creates an actual `gcr.io/foo/bar:_` image.

What should we be using in test goals?

Reply:
@antoineco it's a hack. Earlier specifying `--no-push` on kaniko would not create a tag on gcr yet it would update the cache layers which was perfect. Something changed recently and the cloudbuild stated erroring out with authentication errors when  `--no-push` was being specified. Pushing the image with the `_` tag is my hack to update the cache to speed up subsequent builds.

Reply:
OK thanks! I just added this as a comment in the Makefile so we don't forget why we do things that way.